rfcs/0000-snark-batching.md

```diff
@@ -0,0 +1,77 @@
+## Summary
+[summary]: #summary
+
+Complicate the SNARK pool so that it can do "batch verification" - verification of several proofs at once.
+
+## Motivation
+[motivation]: #motivation
+
+Verification of new SNARK work becomes a throughput bottleneck at high transaction-per-second configurations (saith [#3981](https://github.com/CodaProtocol/coda/pull/3981)). There is a more efficient "batch verification" that we can use - by verifying N SNARKs at once, we spend `600ms + N*10ms` instead of `N * 600ms` <sup>(citation needed)</sup>. This comes with a catch: if batch verification fails, we only know that _some_ SNARK in there failed to verify, but not which. In the case that a batch fails to verify, we need to do extra work to attribute the failure to a particular sender IP / prover, and know which of the SNARKs we can actually include in the pool. By sending incorrect or fraudulent proofs, an attacker can degrade the throughput of a block producer. Each transaction in the block needs to come with two (or occasionally one) proofs, and if not enough proofs are available, the block producer is forced into leaving profit on the table.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+The proposed mechanism is to have two "batch modes": Batch All and Batch by Prover, and to split provers into two categories: Trusted and Untrusted.
```

Comment:
Do you think batching snarks that are required immediately (if the block producer is scheduled to create the next block) would also be nice?  if we wait too long to batch verify/ or some snarks are in a batch that won't be verified until after the diff is created then we might not be able include as many transactions if they were serialized. But this might not be as beneficial from a performance standpoint. 
Also, we have this `delay` factor in scan state so, if there are enough provers and the coordinator picks the work that are needed first then the nodes producing blocks should have the required work at least a block before they need it (if the delay is at least 1). So maybe it is ok

Reply:
@deepthiskumar how can we detect which of these proofs are required in the next block? I think it makes sense to eagerly start a batch before the timeout/max batch size if there's a pending block production deadline, but I'm not sure how to know which proofs to give that treatment.

Reply:
You can query the scan state at the best tip to get all the required work for the next block. `work_statements_for_new_diff` in `transaction_snark_scan_state.ml` will fetch you that